src/lib/InstantSearch.ts

```diff
@@ -494,12 +509,55 @@ See ${createDocumentationLink({
     }
   }
 
-  public onStateChange = () => {
+  public setUiState: (
+    uiState: UiState | ((uiState: UiState) => UiState)
+  ) => void = uiState => {
+    if (!this.mainHelper) {
+      throw new Error(
+        withUsage('The `start` method needs to be called before `setUiState`.')
+      );
+    }
+
+    // We refresh the index UI state to update the local UI state that the
+    // main index passes to the function form of `setUiState`.
+    this.mainIndex.refreshUiState();
+    const nextUiState =
+      typeof uiState === 'function'
+        ? uiState(this.mainIndex.getWidgetState({}))
+        : uiState;
+
+    const setIndexHelperState = (indexWidget: Index) => {
+      if (__DEV__) {
+        checkIndexUiState({
+          index: indexWidget,
+          indexUiState: nextUiState[indexWidget.getIndexId()],
+        });
+      }
+
+      indexWidget.getHelper()!.overrideStateWithoutTriggeringChangeEvent(
+        indexWidget.getWidgetSearchParameters(indexWidget.getHelper()!.state, {
+          uiState: nextUiState[indexWidget.getIndexName()],
+        })
+      );
+
+      indexWidget
+        .getWidgets()
+        .filter(isIndexWidget)
+        .forEach(setIndexHelperState);
+    };
+
+    setIndexHelperState(this.mainIndex);
+
+    this.scheduleSearch();
+    this.onInternalStateChange();
+  };
+
+  public onInternalStateChange = () => {
     const nextUiState = this.mainIndex.getWidgetState({});
 
     this.middleware.forEach(m => {
       m.onStateChange({
-        state: nextUiState,
+        uiState: nextUiState,
```

Comment:
Not sure if that matters since middleware is not yet documented, but it's a breaking change I think

Reply:
It's not documented and under an `EXPERIMENTAL_` prefix.